File: ts/test/remap.test.ts
// Copyright 2026 Wanjia He
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Topic remap tests. Mirrors go/remap_test.go.
//
// The TS SDK has no Writer, so these are driven off the Go-produced
// multi-group-mixed fixture (topics: "/groupA/data", "/groupB/x",
// "/groupB/y"). We verify that remap only rewrites exposed names without
// changing message order/payloads, that topicNames filtering and sample()
// operate in exposed-name space, and that colliding remaps throw.

import { createHash } from "node:crypto";

import { describe, expect, it } from "vitest";

import { Reader, TopicRemapError, type SampleQuery } from "../src/index.js";

import { BytesReadSource, readFixture } from "./helpers.js";

const FIXTURE = "multi-group-mixed.td";
const REMAP = { "/groupA/data": "/groupA/data_v2" };

function hex(data: Uint8Array): string {
  return createHash("sha256").update(data).digest("hex");
}

interface Row {
  ts: bigint;
  name: string;
  data: string;
}

async function collect(reader: Reader, opts = {}): Promise<Row[]> {
  const out: Row[] = [];
  for await (const msg of reader.readMessages({ copy: true, ...opts })) {
    out.push({ ts: msg.timestamp, name: msg.topicName, data: hex(msg.data) });
  }
  return out;
}

describe("topic remap", () => {
  it("rewrites only exposed names, preserving order and payloads", async () => {
    const bytes = readFixture(FIXTURE);
    const base = await collect(new Reader(new BytesReadSource(bytes)));
    const remapped = await collect(
      new Reader(new BytesReadSource(bytes), { topicRemap: REMAP }),
    );

    expect(remapped.length).toBe(base.length);
    for (let i = 0; i < base.length; i++) {
      const expectedName =
        base[i]!.name === "/groupA/data" ? "/groupA/data_v2" : base[i]!.name;
      expect(remapped[i]!.name).toBe(expectedName);
      expect(remapped[i]!.ts).toBe(base[i]!.ts);
      expect(remapped[i]!.data).toBe(base[i]!.data);
    }
    // The in-file name must no longer appear.
    expect(remapped.some((r) => r.name === "/groupA/data")).toBe(false);
  });

  it("topicNames filter accepts the exposed name", async () => {
    const bytes = readFixture(FIXTURE);
    const reader = new Reader(new BytesReadSource(bytes), { topicRemap: REMAP });
    const rows = await collect(reader, { topicNames: ["/groupA/data_v2"] });

    expect(rows.length).toBeGreaterThan(0);
    expect(rows.every((r) => r.name === "/groupA/data_v2")).toBe(true);

    // An exposed name that maps to nothing matches nothing.
    const reader2 = new Reader(new BytesReadSource(bytes), {
      topicRemap: REMAP,
    });
    const none = await collect(reader2, { topicNames: ["/nope"] });
    expect(none.length).toBe(0);
  });

  it("summary reports exposed names", async () => {
    const bytes = readFixture(FIXTURE);
    const reader = new Reader(new BytesReadSource(bytes), { topicRemap: REMAP });
    const summary = await reader.summary();
    const names = new Set<string>();
    for (const ti of summary.topicsInfos) {
      for (const tm of ti.topicMetadatas) {
        names.add(tm.name);
      }
    }
    expect(names.has("/groupA/data_v2")).toBe(true);
    expect(names.has("/groupB/x")).toBe(true);
    expect(names.has("/groupA/data")).toBe(false);
  });

  it("throws on a colliding remap (lazily, on first use)", async () => {
    const bytes = readFixture(FIXTURE);
    // Rename "/groupB/x" onto the untouched in-file name "/groupB/y".
    const reader = new Reader(new BytesReadSource(bytes), {
      topicRemap: { "/groupB/x": "/groupB/y" },
    });
    await expect(reader.summary()).rejects.toThrow(TopicRemapError);
    await expect(reader.summary()).rejects.toThrow(/duplicate exposed name/);

    const reader2 = new Reader(new BytesReadSource(bytes), {
      topicRemap: { "/groupB/x": "/groupB/y" },
    });
    await expect(
      (async () => {
        for await (const _ of reader2.readMessages()) {
          // drain
        }
      })(),
    ).rejects.toThrow(TopicRemapError);
  });

  it("sample() accepts exposed names and matches the in-file result", async () => {
    const bytes = readFixture(FIXTURE);
    // Ground truth: read /groupA/data messages with no remap.
    const base = await collect(new Reader(new BytesReadSource(bytes)));
    const first = base.find((r) => r.name === "/groupA/data");
    expect(first).toBeDefined();

    const reader = new Reader(new BytesReadSource(bytes), { topicRemap: REMAP });
    const query: SampleQuery = {
      topic: "/groupA/data_v2",
      timestamps: [first!.ts],
    };
    const out = await reader.sample([query]);
    expect(out[0]!.length).toBe(1);
    expect(out[0]![0]!.found).toBe(true);
    expect(out[0]![0]!.timestamp).toBe(first!.ts);
    expect(hex(out[0]![0]!.data)).toBe(first!.data);
  });

  it("no remap is identity", async () => {
    const bytes = readFixture(FIXTURE);
    const base = await collect(new Reader(new BytesReadSource(bytes)));
    const identity = await collect(
      new Reader(new BytesReadSource(bytes), { topicRemap: {} }),
    );
    expect(identity).toEqual(base);
  });
});

File: go/turbodata/remap_test.go
// Copyright 2026 Wanjia He
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package turbodata

import (
	"bytes"
	"strings"
	"testing"
)

// buildRemapReader builds a two-topic file ("a", "b") and returns a Reader
// over it with the given remap applied.
func buildRemapReader(t *testing.T, remap map[string]string) *Reader {
	t.Helper()
	raw := buildFile(t, func(w *Writer) {
		mustOpenTopics(t, w, []string{"a", "b"}, []map[string]any{{}, {}})
		mustWriteMessage(t, w, "a", []byte("a1"), 10)
		mustWriteMessage(t, w, "b", []byte("b1"), 20)
		mustWriteMessage(t, w, "a", []byte("a2"), 30)
		mustWriteMessage(t, w, "b", []byte("b2"), 40)
		mustCloseTopic(t, w)
		mustClose(t, w)
	})
	return NewReader(bytes.NewReader(raw), WithTopicRemap(remap))
}

func TestReaderRemapOutputNames(t *testing.T) {
	r := buildRemapReader(t, map[string]string{"a": "a_v2"})
	msgs := collect(t, r)

	want := []testMsg{
		{10, "a_v2", []byte("a1")},
		{20, "b", []byte("b1")},
		{30, "a_v2", []byte("a2")},
		{40, "b", []byte("b2")},
	}
	if len(msgs) != len(want) {
		t.Fatalf("message count: got %d want %d", len(msgs), len(want))
	}
	for i := range want {
		if msgs[i].ts != want[i].ts || msgs[i].name != want[i].name || !bytes.Equal(msgs[i].data, want[i].data) {
			t.Errorf("[%d] got (%d,%q,%q) want (%d,%q,%q)",
				i, msgs[i].ts, msgs[i].name, msgs[i].data, want[i].ts, want[i].name, want[i].data)
		}
	}
}

func TestReaderRemapTopicNamesFilter(t *testing.T) {
	r := buildRemapReader(t, map[string]string{"a": "a_v2"})

	// Filtering must accept the exposed name and return only that topic's
	// messages, labeled with the exposed name.
	msgs := collect(t, r, WithTopicNames([]string{"a_v2"}))
	want := []testMsg{
		{10, "a_v2", []byte("a1")},
		{30, "a_v2", []byte("a2")},
	}
	if len(msgs) != len(want) {
		t.Fatalf("message count: got %d want %d (%+v)", len(msgs), len(want), msgs)
	}
	for i := range want {
		if msgs[i].ts != want[i].ts || msgs[i].name != want[i].name || !bytes.Equal(msgs[i].data, want[i].data) {
			t.Errorf("[%d] got (%d,%q,%q) want (%d,%q,%q)",
				i, msgs[i].ts, msgs[i].name, msgs[i].data, want[i].ts, want[i].name, want[i].data)
		}
	}
}

func TestReaderRemapSummaryExposedNames(t *testing.T) {
	r := buildRemapReader(t, map[string]string{"a": "a_v2"})
	summary, err := r.Summary()
	if err != nil {
		t.Fatalf("Summary: %v", err)
	}
	got := make(map[string]bool)
	for _, ti := range summary.TopicsInfos {
		for _, tm := range ti.TopicMetadatas {
			got[tm.Name] = true
		}
	}
	if !got["a_v2"] || !got["b"] {
		t.Errorf("summary names: got %v want a_v2 and b", got)
	}
	if got["a"] {
		t.Errorf("summary still exposes in-file name %q", "a")
	}
}

func TestReaderRemapCollisionError(t *testing.T) {
	// Renaming "a" onto the untouched in-file name "b" collapses two topics
	// onto one exposed name and must error on first use.
	r := buildRemapReader(t, map[string]string{"a": "b"})

	if _, err := r.Summary(); err == nil {
		t.Errorf("Summary: expected collision error, got nil")
	} else if !strings.Contains(err.Error(), "duplicate exposed name") {
		t.Errorf("Summary: error %q should mention duplicate exposed name", err)
	}

	if _, err := r.ReadMessages(); err == nil {
		t.Errorf("ReadMessages: expected collision error, got nil")
	}
}

func TestReaderRemapSample(t *testing.T) {
	r := buildRemapReader(t, map[string]string{"a": "a_v2"})

	out, err := r.Sample([]SampleQuery{
		{Topic: "a_v2", Timestamps: []int64{10, 25, 30}},
	})
	if err != nil {
		t.Fatalf("Sample: %v", err)
	}
	want := []SampleResult{
		{Found: true, Timestamp: 10, Data: []byte("a1")},
		{Found: true, Timestamp: 10, Data: []byte("a1")},
		{Found: true, Timestamp: 30, Data: []byte("a2")},
	}
	assertResultsEqual(t, out[0], want)
}
